index.js

```diff
@@ -159,6 +161,10 @@ class S3Adapter {
   getFileLocation(config, filename) {
     const fileName = filename.split('/').map(encodeURIComponent).join('/');
     if (this._directAccess) {
+      if (this._presignedUrl) {
```

Comment:
Thanks for the PR. It looks great. Could you provide an example of url that aws generates? I'm wondering if this option can be used combined with baseUrl option to return a presignedUrl over a custom domain.

Reply:
https://<BUCKET_NAME>.s3.amazonaws.com/<FILE_NAME>?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=<AWS_ACCESS_KEY_ID>%2F20201117%2F<AWS_REGION>%2Fs3%2Faws4_request&X-Amz-Date=20201117T132727Z&X-Amz-Expires=30&X-Amz-Signature=0b6389431cc639f2f1694815f2b86fd855294076c0d56587fb17bf74fe2f9a9b&X-Amz-SignedHeaders=host

Reply:
So in this case I believe it would be possible to make the new options to work with baseUrl, right? Could you please change that in the PR?

Reply:
I don't know if I understood correctly, but it still works with `baseUrl`.
The purpose of this change is to use the temporary URL feature of AWS S3, so it has to be returned.
If you wanted to use `baseUrl`, just do not configure `presignedUrl` and do whatever treatment you want "at the other end", using "a presigned url" or not.

Reply:
Is it possible to return the Parse Server domain for that URL if `directAccess` is set to false? The adapter would internally translate to the S3 URL and therefore hide the S3 URL from the end user.

Reply:
First, according to current logic if `directAccess` is set to `false` we need to return the parse URL without modifications. The URL only changes if `directAccess` is configured for `true`.

Now see the issue (https://github.com/parse-community/parse-server-s3-adapter/issues/78) text:
> This is a feature suggestion/request. It would be nice to be able to retrieve a presigned URL with dynamic expiration from the Amazon S3 bucket when directAccess is enabled. This would allow more security by allowing us to lock down the S3 bucket with public access turned off, and still provide the benefit of not proxy'ing the download through the parse server.

The idea here is precisely to remove Parse's responsibility to mediate the file download, but to keep the file safe.
The reason is that today it is only possible to return the S3 URL directly if the files are public, since with this change they can be made private.

Two more points:
1. I may be mistaken, but the `getFileData(filename)` method only passes the filename as an argument, so it would not be possible to get the other parameters of the URL or header.
2. In theory if you want a temporary URL using only Parse for this, it is already possible using `baseUrl` as a function. You can return something like this: `https://<BASE_URL_WITH_TOKEN_IN_THE_PATH>/<FILE_NAME>`.

Reply:
That's not true. `baseUrl` can be used together with `directAccess`. I can have a bucket in S3 called for example mybucket.com and I can set dns cname from mybucket.com to the S3 bucket dns. When using `directAccess` and `baseUrl` together for this case, the link will be generated and the file will be downloaded directly from S3 but using mybucket.com instead of aws dns. It would be good if the same process could happen also with the presigned url. It is a simples change. I can do it myself if you prefer.

Reply:
I understand your point and really you are correct. I made the change that I believe will leave things as you expect.
However there is a detail that can be a problem. Assuming that `directAccess`, `presignedUrl` and `baseUrlDirect` were set to `true`, the final URL will not contain `bucketPrefix`, so it will not work as expected for the DNS example.

Reply:
Could you please try with something like this?

```js
getFileLocation(config, filename) {
    const fileName = filename.split('/').map(encodeURIComponent).join('/');
    if (this._directAccess) {
      const fileKey = `${this._baseUrlDirect ? '' : this._bucketPrefix}${fileName}`;

      let presignedUrl = '';
      if (this._presignedUrl) {
        const params = {
          Bucket: this._bucket,
          Key: fileKey,
          Expires: this._presignedUrlExpires,
        };
        presignedUrl = this._s3Client.getSignedUrl('getObject', params);
      }
      if (this._baseUrl) {
        let directAccessUrl;

        if (typeof this._baseUrl === 'function') {
          directAccessUrl = `${this._baseUrl(config, filename)}/${fileKey}`;
        } else {
          directAccessUrl = `${this._baseUrl}/${fileKey}`;
        }

        if (this._presignedUrl) {
          directAccessUrl += presignedUrl.substring(presignedUrl.indexOf('?'));
        }
        return directAccessUrl;
      }
      if (this._presignedUrl) {
        return presignedUrl;
      }
      return `https://${this._bucket}.s3.amazonaws.com/${fileKey}`;
    }
    return `${config.mount}/files/${config.applicationId}/${fileName}`;
  }
```

I believe it will fix the problem you mentioned and reduce the code complexity.

Reply:
The complexity of the code actually decreases, but it hasn't solved the problem yet.
However, this implementation has 2 more problems if `baseDirectUrl` is set to `true`:
1. `this._s3Client.getSignedUrl()` will not consider the `bucketPrefix`.
2. The AWS direct url will also not consider `bucketPrefix`.

I liked your suggestion and I know how to solve the 2 problems I mentioned here, but the initial problem will continue.

Would it be the case to add a disclaimer in `README.md` or do you have any other suggestions?

Reply:
I believe that's actually the idea of the baseDirectUrl parameter: https://github.com/parse-community/parse-server-s3-adapter/pull/9

Reply:
For public URL yes, but not for AWS URL.
In the PR you linked, if `baseUrl` does not exist, the AWS URL is not affected by the `baseUrlDirect` parameter.

In fact, the problem I mentioned based on your DNS example already existed. Whoever wanted to use this approach with DNS, could not set `baseDirectUrl` to `true`. I'm sorry I didn't notice that before 😞.

So, I will improve the code based on your suggestion and I think we can merge this PR without further changes.